src/main/java/org/spongepowered/common/mixin/core/entity/item/MixinEntityMinecartTNT.java

```diff
@@ -24,10 +24,57 @@
  */
 package org.spongepowered.common.mixin.core.entity.item;
 
+import com.google.common.collect.Lists;
+import net.minecraft.entity.Entity;
 import net.minecraft.entity.item.EntityMinecartTNT;
+import net.minecraft.entity.projectile.EntityArrow;
 import org.spongepowered.api.entity.vehicle.minecart.MinecartTNT;
+import org.spongepowered.api.event.SpongeEventFactory;
+import org.spongepowered.api.event.cause.Cause;
+import org.spongepowered.api.event.cause.NamedCause;
+import org.spongepowered.api.event.world.ExplosionEvent;
+import org.spongepowered.api.world.World;
 import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+import org.spongepowered.asm.mixin.injection.callback.LocalCapture;
+import org.spongepowered.common.SpongeImpl;
+
+import java.util.List;
+
+import javax.annotation.Nullable;
 
 @Mixin(EntityMinecartTNT.class)
 public abstract class MixinEntityMinecartTNT extends MixinEntityMinecart implements MinecartTNT {
+
+    @Nullable private Entity shooter;
+
+    @Inject(method = "attackEntityFrom(Lnet/minecraft/util/DamageSource;F)Z", at = @At(value = "INVOKE", target =
+            "Lnet/minecraft/entity/item/EntityMinecartTNT;explodeCart(D)V", shift = At.Shift.BEFORE), locals = LocalCapture.CAPTURE_FAILEXCEPTION)
+    public void processAttackEntityFromCauseCapture(CallbackInfo ci, EntityArrow arrow) {
+        this.shooter = arrow;
+    }
+
+    @Inject(method = "explodeCart(D)V", at = @At(value = "INVOKE", target = "Lnet/minecraft/world/World;createExplosion(Lnet/minecraft/entity/Entity;DDDFZ)"
+            + "Lnet/minecraft/world/Explosion;", shift = At.Shift.BEFORE), cancellable = true)
+    public void processExplosion(CallbackInfo ci) {
+        List<Object> cause = Lists.newArrayList();
+
+        if (this.shooter != null) {
+            cause.add(NamedCause.igniter(this.shooter));
+        }
+
+        cause.add(NamedCause.source(this));
+
+        // Reset state
+        this.shooter = null;
+
+        ExplosionEvent.Pre event = SpongeEventFactory.createExplosionEventPre(Cause.of(cause), (World) this.worldObj);
+        SpongeImpl.postEvent(event);
```

Comment:
postEvent returns cancellation status. You can inline it (and I'd prefer it for leaner mixins).